instrumentation/httpd/src/otel/mod_otel.cpp

```diff
@@ -0,0 +1,387 @@
+  auto ExtraRequestDataBuffer = apr_palloc(r->pool, sizeof(ExtraRequestData));
+  // and use placement new
+  req_data_out = new (ExtraRequestDataBuffer) ExtraRequestData;
+
+  apr_pool_userdata_setn(ExtraRequestDataBuffer, kOpenTelemetryKeyOutboundNote,
+                         (apr_status_t(*)(void *))ExtraRequestData::Destruct, req->pool);
+
+  opentelemetry::trace::StartSpanOptions startOpts;
+  startOpts.kind = opentelemetry::trace::SpanKind::kClient;
+  startOpts.parent = req_data->span->GetContext();
+  auto span = get_tracer()->StartSpan(kSpanNamePrefix + req->method, startOpts);
+
+  req_data_out->span  = span;
+  HttpdStartSpanAttributes startAttrs;
+  startAttrs.server_name = r->server->server_hostname;
+  startAttrs.method = req->method;
+  startAttrs.url = req->filename;
+  if (startAttrs.url.substr(0,6) == "proxy:")
+  {
+    startAttrs.url = startAttrs.url.substr(6);
+  }
+
+  req_data_out->StartSpan(startAttrs);
+
+  auto scope = get_tracer()->WithActiveSpan(span);
+
+  // mod_proxy simply copies request headers from client therefore inject is into headers_in
+  // instead of headers_out
+  switch (config.propagation)
+  {
+    case OtelPropagation::TRACE_CONTEXT:
+      PropagatorTraceContext.Inject(HttpdSetter, *req->headers_in, opentelemetry::context::RuntimeContext::GetCurrent());
+      break;
+    case OtelPropagation::B3_SINGLE_HEADER:
+      PropagatorB3SingleHeader.Inject(HttpdSetter, *req->headers_in, opentelemetry::context::RuntimeContext::GetCurrent());
+      break;
+    case OtelPropagation::B3_MULTI_HEADER:
+      PropagatorB3MultiHeader.Inject(HttpdSetter, *req->headers_in, opentelemetry::context::RuntimeContext::GetCurrent());
+      break;
+    default:  // suppress warning
+      break;
+  }
+
+  return DECLINED;
+}
+
+static int proxy_end_handler(int *status, request_rec *r)
+{
+  request_rec *req = r->main ? r->main : r;
+
+  ExtraRequestData *req_data_out;
+  apr_pool_userdata_get((void **)&req_data_out, kOpenTelemetryKeyOutboundNote, req->pool);
+
+  if (!req_data_out)
+  {
+    return DECLINED;
+  }
+
+  int st_code = (status && *status) ? *status:r->status;
+  char *proxy_error = apr_table_get(r->notes, "error-notes");
+  if (proxy_error)
+  {
+    req_data_out->span->SetStatus(opentelemetry::trace::StatusCode::kError, proxy_error);
+  }
+
+  // finish span
+  req_data_out->EndSpan({
+    st_code,
+    req->bytes_sent
+  });
+  req_data_out->Destruct(req_data_out);
+
+  return DECLINED;
+}
+
+/////////////////////////////////////////////////
+// PARSING CONFIGURATION OPTIONS
+/////////////////////////////////////////////////
+
+static const char *otel_set_exporter(cmd_parms *cmd, void *cfg, const char *arg)
+{
+  if (!strcasecmp(arg, "file"))
+    config.type = OtelExporterType::OSTREAM;
+  else if (!strcasecmp(arg, "otlp"))
+    config.type = OtelExporterType::OTLP;
+  else
+    return "Unknown exporter type - can be file or otlp";
+
+  return NULL;
+}
```

Comment:
Please correct me if I'm wrong here, but I  think any kind of instrumentation shouldn't define an exporter. Exporters should only be defined by the end-user application developer. Otherwise you could only use _one_ instrumentation at a time (e. g. Apache), but never two or more (e. g. Apache and mysql) as they may conflict (e. g. intitialize two OTEL exporters).

As  rule of thumb, instrumentation should only use the OpenTelemetry API, but not the SDK.

Reply:
Also, see for example who Go provides instrumentation for its standard HTTP server: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/example/server/server.go

The exporter is configurable via standard SDK methods, and the instrumentation is agnostic of it.

Reply:
Just realized that in case of Apache (and nginx) we're not actually instrumenting a _library_, but and _end-user application_.

So it might be ok to specify an exporter in this case. However, it's still a bit troubling for me that user can't specify their custom exporters or configure exporters/processors without recompiling.